documentation/SOFTWARE_DESIGN.md

```diff
@@ -65,7 +65,9 @@ Sokoban constructor(const vec<vec<string>> levels)
 // class properties
 const vec<vec<string>> levels, read-only
 uint current_level // index into levels
-uint moves
+vector<Direction> moves // history of moves made by the player
+unit player_x // the player's column position
+unit player_y // the player's row position
```

Comment:
I suggest creating a class called Point with attributes uint x and y.

Reply:
`Point` is totally possible, and I did see that in your code, but on the other hand, I don't see a super strong motivation for using it either and it should be an easy refactor later on if need be. What is your goal/motivation in using a `Point` class, or what problem are we solving by adding abstraction here besides neatly packing both vars into one entity (since they are indeed related)?

In my original Sokoban implementation, separate `x`/`y` values worked OK: https://github.com/ggorlen/sokoban/blob/master/js/sokoban.js#L36

To set the `x` and `y` values (or populate `Point`/`Player`/`Position` (whatever we call it)), I figure we'll scan the board when a level loads and locate the `@` symbol, then manipulate these coordinates whenever `move(direction)` is called, which is always performed relative to whatever `x`/`y` is. The client can never touch the `x`/`y` values directly -- it's internal state and should be pretty easy to refactor later on if need be.

BTW, `uint` means "unsigned integer", so it's not a typo of `unit`. Sorry if that wasn't clear!

Reply:
I am not sure yet. But what I am thinking is if I need to check multiple points and use that as a parameter. I can just have func(Point1, Point2) rather than func(x1, y1, x2, y2). It might get crowded to pass as a parameter. We might need this down the line. It might also be handy when we are handling comparing coordinates for later. Also if we decide to have a 3 dimensional. A point can be easily changed by adding another attribute z = 0. I don't know. Maybe we don't need it. We'll see as we go along.

I agree that x/y values have to be private or internal.

What is a `unit`? I am familiar with it.

Reply:
Re; `uint`, I apologize -- that's a typo on my part. I pushed a commit for `uint`, the intended type.

We could do a class, but I don't anticipate that we'll need to pass `player.x` and `player.y` as the parameter to any/many functions (I didn't in my implementation). It should only be read internally by `Sokoban` functions as an implementation detail.

In your code, I noticed you were passing `Point`s into many functions, which moves control of the player (a part of the internal state of a Sokoban game) to the client. But if the client is stuck managing this state and making sure it doesn't get out of sync with the board (the `'@'` symbol is the true player position), it winds up having to track a lot of Sokoban logic it shouldn't have to worry about. Suddenly, the Sokoban class doesn't entirely implement Sokoban -- a combination of the client and Sokoban then implement Sokoban, so the abstraction seems leaky and incomplete to me rather than a black box the client can "just use" to play the game without the possibility of winding up in an illegal position or violating the (now complex) contract between `Sokoban` and the client of the class.

At the end of the day, `Point` or no `Point` should be pretty minor because it shouldn't be much of a big player in the code, just a pair of vars that are read whenever we need to know where the player is as a convenience (and possibly premature optimization) so we don't have to do a linear search on every move to figure out where the `'@'` character is (although that'd technically work). Thinking of it this way might help explain why it's not terribly important data -- it redundantly aliases a piece of info the 2d board vector, which is the critical piece of game state.

Does this design make sense? I think it's a pretty important point, and I admit I lost the thread on your draft PR when `Point` showed up and began to be used as a parameter to many different functions. That's a pretty different approach than what I'm advocating for here, so I think we'd need good justification for why the player position shouldn't be controlled by the core game engine.

The purpose of this file is to agree on the complete public interface for Sokoban which we'll be implementing without alteration (hopefully!). So it's a good time to experiement and have these conversations, because once we do, that'll be the contract that the clients (WASM glue code and doctest suite) will use to play the game, and we probably won't want to mess with it until we're done with the basic game.

Anyway, let's continue the conversation and hopefully this helps clarify my thoughts about some of the ideas you brought up in your draft.

Reply:
> In your code, I noticed you were passing `Point`s into many functions, which moves control of the player (a part of the internal state of a Sokoban game) to the client. But if the client is stuck managing this state and making sure it doesn't get out of sync with the board (the `'@'` symbol is the true player position), it winds up having to track a lot of Sokoban logic it shouldn't have to worry about. 

In my code, I thought I was passing around Points just to read the positions of a player or any element on the board. I can't see yet how the client would be able to mess around it. The helper functions that I created will be private. Will that solve the issue?

> Does this design make sense? I think it's a pretty important point, and I admit I lost the thread on your draft PR when `Point` showed up and began to be used as a parameter to many different functions. That's a pretty different approach than what I'm advocating for here, so I think we'd need good justification for why the player position shouldn't be completely controlled by the core game engine as an internal implementation detail (that's partially why `Point` vs `px`/`py` is an easy refactor and shouldn't matter much -- it's not part of the public interface).

I agree. the player position should be completely controlled by the core game. In my code, I have two segments. One implements most of the methods in the Design document. I intend the bottom segment to be private and used as a helper function for move(). I hope that my code is not completely dismissed because it has a different approach. Maybe I need more guidance on how I can improve it. I definitely need this feedback and maybe help me identify which functions would potentially create abstraction leak.

Reply:
I see, thanks for clarifying. Since they're class variables, there's no need for them to be parameters though, any function in the class can read/write them.

Reply:
True. Now that we added the player positions, I can access that directly.